### PR TITLE
Namespace generators

### PR DESCRIPTION
**Pull request by RaeRachael**
- routes are added correctly into config/routes

- files in controllers and dashboards appear in the right location

- created a list for tables to be excluded from being made into routes/dashboards (excluded tables don't created console warning messages

### CODE REVIEW

File: spec/generators/routes_generator_spec.rb [hunk 1]
```diff
@@ -120,6 +115,7 @@ class ModelWithoutDBTable < ApplicationRecord; end
 
     run_generator
 
-    expect(routes).to contain('root to: "customers#index')
+    expect(routes).to contain('root to: "application#show"')
+    # Why does the root get changed when namespace is included..
```
Comment (pablobm) [inline spec/generators/routes_generator_spec.rb:119]: It's actually not changing. Instead, this refers to an additional `root` line in the file, which was there before this PR. The one that we were originally testing for (`customers#index`) is not there any more, so you have found that this other one is there and changed the expectation.

This example runs the generator against an existing, mostly empty routes file. If you search for `application#show` in the codebase, you'll find it:

https://github.com/thoughtbot/administrate/blob/84084235463a6761f25108382cb81c13632d2052/spec/app_templates/config/routes.rb#L1-L3

So that `application#show` is there already, is correct, and shouldn't change when the generator runs. That's fine. When the generator runs, new lines are added. If you insert a `puts File.read(routes)` right after `run_generator`, you'll see what the routes file actually looks like by the time the expectation is checked. It looks like this:

```ruby
Rails.application.routes.draw do
  namespace :admin do
    namespace :active_record do
      resources :internal_metadata
    end
    namespace :active_record do
      resources :schema_migrations
    end
    namespace :primary do
      resources :schema_migrations
    end
    resources :customers
    resources :orders
    namespace :blog do
      resources :posts
    end
    resources :countries
    resources :line_items
    resources :log_entries
    resources :pages
    resources :payments
    resources :products
    resources :product_meta_tags
    resources :series
    
    root to: "active_record/internal_metadata#index"
  end
  root to: "application#show"
end
```

So from here you can see that the problem is more with new namespaced that are being added and probably shouldn't: `active_record` and `primary`. The root is assigned arbitrarily to the `#index` of the first model in the list, which happens to be `active_record/internal_metadata`.

I think that this spec example should change so that the expectation is a bit more strict about what's in the file. Perhaps it should match a full file, instead of searching for a specific line.

**Review by pablobm — CHANGES_REQUESTED** [review]

File: spec/generators/routes_generator_spec.rb (hunk 1 above)
Comment (RaeRachael) [reply spec/generators/routes_generator_spec.rb:119]: Thank you for the information here @pablobm,

`puts File.read(routes)` is a helpful visibility idea - slightly annoyed I didn't see it. 

Yeah those namespaces shouldn't be used - so that's helpful to know what to investigate. Also testing the whole file makes more sense.

File: lib/generators/administrate/routes/routes_generator.rb [hunk 2]
```diff
@@ -59,15 +57,32 @@ def database_models
...
+        end
+      end
+
+      def generate_nested_resource_routes(items, index = 0)
+        if index == items.size - ONE
+          "resources :#{items[index]}"
+        else
+          resource_indent = " " * ((index + TWO) * DEFAULT_INDENT)
+          resource = generate_nested_resource_routes(items, index + ONE)
+          indent_resource =  "#{resource_indent}#{resource}"
```
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb:82]: Honestly I can't believe how bad my whitespace appearance is...

File: lib/administrate/constants.rb [hunk 3]
```diff
@@ -0,0 +1,3 @@
+DEFAULT_INDENT = 2
+ONE = 1
+TWO = 2
```
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/constants.rb:3]: Yeah I wasn't sure how to handle this, listing the constants at the top of the `routes generator` file seemed odd. But yeah having them top level object is bad - something I will try to avoid, 

Also not sure if literal `1` and `2` would be ok...

[the remark this answers is not in the document]
Comment (pablobm) [reply lib/administrate/constants.rb:3]: Just to clarify: the constants should go inside the module, not just at the top of the file. Otherwise they'll also be assigned to `Object`.

As for the literals... yes, they would be ok :-) (But if you can find a good name for the `TWO` one, that'd be good too).

File: spec/generators/routes_generator_spec.rb [hunk 4]
```diff
@@ -120,6 +115,6 @@ class ModelWithoutDBTable < ApplicationRecord; end
 
     run_generator
 
-    expect(routes).to contain('root to: "customers#index')
+    expect(routes).to contain('root to: "customers#index"')
```
Comment (pablobm) [inline spec/generators/routes_generator_spec.rb:118]: Good catch :+1:

**Review by pablobm — COMMENTED** [review]
Getting there one step at a time :-)

File: lib/administrate/generator_helpers.rb [hunk 5]
```diff
@@ -1,5 +1,17 @@
+require "find"
+
 module Administrate
   module GeneratorHelpers
+    DEFAULT_INDENT = 2
+    ONE_INDENT = 1
+    TWO_INDENT = 2
```
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/generator_helpers.rb:7]: It is being used as the standard unit of indent, perhaps naming it as `STARDARD_INDENT` would be better?

File: lib/administrate/generator_helpers.rb [hunk 6]
```diff
@@ -1,5 +1,17 @@
...
+    OFFSET = 1
+    INVALID_DATABASE_MODELS_LIST = [
+      "ActiveRecord::SchemaMigration",
+      "ActiveRecord::InternalMetadata",
+      "primary::SchemaMigration",
+    ]
```
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/generator_helpers.rb:13]: I didn't know they are used there - the list here leads them not to be used in the `routes_genator` so that they aren't treated as valid models.

File: lib/administrate/generator_helpers.rb [hunk 7]
```diff
@@ -1,5 +1,17 @@
+require "find"
+
 module Administrate
   module GeneratorHelpers
+    DEFAULT_INDENT = 2
+    ONE_INDENT = 1
+    TWO_INDENT = 2
+    OFFSET = 1
```
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/generator_helpers.rb:8]: that makes sense, will change it to a literal

File: lib/generators/administrate/dashboard/dashboard_generator.rb [hunk 8]
```diff
@@ -122,6 +126,20 @@ def klass
         @klass ||= Object.const_get(class_name)
       end
 
+      def class_path_filename_fragment
+        if @class_path.present?
+          "/#{@class_path.join('/')}"
+        end
+      end
```
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/dashboard/dashboard_generator.rb:133]: Will look into making a test for it

File: lib/administrate/generator_helpers.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/generator_helpers.rb:13]: So I think moving this list into `lib/generators/administrate/routes/routes_generator.rb` would be the best solution

[the remark this answers is not in the document]
Comment (pablobm) [reply lib/administrate/generator_helpers.rb:13]: My mistake, I meant `lib/generators/administrate/routes/routes_generator.rb`

File: lib/administrate/generator_helpers.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (pablobm) [reply lib/administrate/generator_helpers.rb:7]: Oh, I see. I had misunderstood it. The name is ok then (although `STANDARD_INDENT`) also works.

I was a bit confused as to what `ONE_INDENT` and `TWO_INDENT` meant. I get them now. I think I would have implemented this differently, in a way that avoids the index wrangling and removes the need for these two constants. The main idea would be not to send the whole list on to the next recursive method call, but only the bits of the list that remain to be rendered. This is a common pattern in functional programming. Let me have a try...

[the remark this answers is not in the document]
Comment (pablobm) [reply lib/administrate/generator_helpers.rb:7]: Actually, before I try to reimplement this, I found a new bug. If there is more than one namespaced resource, each will get its own `namespace` entry, like this:
```ruby
    namespace :blog do
      resources :posts
    end
    namespace :blog do
      resources :users
    end
```

To solve this, I suggest that `generate_resource_routes` takes the whole list of resources, instead of one at a time. Then it'll have to change to handle this list.

[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/generator_helpers.rb:7]: That makes sense, we'll look into this tomorrow. Just for my clarity would this break any functionality or is it just badly formatted?

[the remark this answers is not in the document]
Comment (pablobm) [reply lib/administrate/generator_helpers.rb:7]: It should technically work; it's just unexpected (and arguably bad) formatting.

File: lib/administrate/generator_helpers.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/generator_helpers.rb]: Implementation to fix the, 
    ```
    namespace :blog do
      resources :posts
    end
    namespace :blog do
      resources :users
    end
```
removed the offset approach

File: lib/administrate/generator_helpers.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/generator_helpers.rb]: should be fixed, any advice for a good way to test this - as the current system in the `routes_generator_spec` uses the example app models (which don't have extended namespace structure.

File: lib/administrate/generator_helpers.rb (hunk 6 above)
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/administrate/generator_helpers.rb]: moved

File: lib/generators/administrate/routes/routes_generator.rb [hunk 9]
```diff
@@ -50,30 +53,92 @@ def dashboard_resources
...
+          ]
+        end
+        output
+      end
+
+      def add_namespace_to_hash(split_path, current_state)
+        if split_path.length > 1
+          current_state[split_path[0]] ||= [{}, []]
+          current_namespace = current_state[split_path[0]]
+          current_namespace = [
```
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb:72]: used to fix a line length issue..

File: lib/generators/administrate/routes/routes_generator.rb [hunk 10]
```diff
@@ -50,30 +53,92 @@ def dashboard_resources
...
+        output = [{}, []]
+        dashboard_resources.each do |resource_path|
+          split_path = resource_path.split("/")
+          output = [
+            add_namespace_to_hash(split_path, output[0]),
+            add_resources_to_hash(split_path, output[1]),
+          ]
+        end
+        output
+      end
```
Comment (pablobm) [inline lib/generators/administrate/routes/routes_generator.rb:66]: I see what you are trying to do there. It's something complicated to code and test, so I can't complain about your implementation. What I'm concerned is that, precisely because it's difficult to test, it probably doesn't cover as many cases as we think.

Here are a couple (asking without actually checking, perhaps you have this covered). Does it support...:

* Multiple levels of nesting?
* Resources with nesting? Example:

```ruby
resources :authors do # NOTE: not `namespace`
  resources :posts
end
# This gives you things like:
#  - /authors
#  - /authors/:id
#  - /authors/:id/posts
#  - /authors/:id/posts/:id
```

Testing this is difficult with our current spec, which is a bit more end-to-end testing the functionality of the generator. If we wanted to test these edge cases (which I think we should), we would end up complicating that spec too much.

Instead, I think this "figuring out the tree" functionality should be moved to a separate class that can be unit-tested for edge cases, leaving a general example for this spec. It could have an interface like this:

```ruby
resources = [["blog/authors"], ["blog/posts"], ["countries"], ["customers"]]
rtc = RouteTreeCalculator.new(resources) # First name that came to mind; not attached to it
rtc.tree # => Not sure exactly what, but something that `generate_resource_routes` will consume
```

Does this make sense?

File: lib/generators/administrate/dashboard/dashboard_generator.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (pablobm) [reply lib/generators/administrate/dashboard/dashboard_generator.rb:133]: This might also benefit from my advice about factoring out a difficult algorithm into a separate class, and then unit-testing the separate class. That way the generator spec doesn't need to get too complicated, and you still get to test edge cases.

**Review by pablobm — COMMENTED** [review]
Sorry to be that person :disappointed: This is trickier than any of us thought at first sight, and it's only apparent after digging into it. You are putting a great effort, and I'm sure you can push it through the finish line.

File: lib/generators/administrate/routes/routes_generator.rb (hunk 10 above)
Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb:66]: The hash output works well (I wrote it in a separate folder along with some tests) I think moving the function out into new folder in administrate will make the testing and code clearer.

I am confident it leads to the correct nesting structure - though I labeled them all as a namespace instead of a resource so that would need to change? or we need to figure out when to use namespace and not resource..

File: lib/generators/administrate/routes/routes_generator.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb:72]: It is used in the declaration, lines 73 and 74

File: lib/generators/administrate/routes/templates/routes.rb.erb [hunk 11]
```diff
@@ -1,5 +1,4 @@
 namespace :<%= namespace %> do
-<% dashboard_resources.each do |resource| %>    resources :<%= resource %>
-<%end%>
-    root to: "<%= dashboard_resources.first %>#index"
-  end
+<%= generate_resource_routes(dashboard_resources_hash) %>
```
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/routes/templates/routes.rb.erb:2]: Yeah I can see that!

File: lib/generators/administrate/routes/routes_generator.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (pablobm) [reply lib/generators/administrate/routes/routes_generator.rb:72]: Yes and no:

* Line 71: first definition of `current_namespace`.
* Lines 72-75: second definition, referring back to the previous one.
* Line 76: end of `if` block.
* Line 77: last expression of the method, which only refers to `current_state`.
* Line 78: end of method.

So after its last definition (lines 72-75), `current_namespace` is never used for anything. The method doesn't do anything afterwards that makes use of it, and I don't think it takes part in any side-effect.

So this method is equivalent to the following:

```ruby
def add_namespace_to_hash(split_path, current_state)
  if split_path.length > 1
    current_state[split_path[0]] ||= [{}, []]
  end
  current_state
end
```

[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb:72]: 73, 74. add to the `current namespace` which is current done by passing in the `current namespace element` to be added to . (I can see why the linter is unhappy, but not sure if the process fits what the linter is unhappy about) I only put in the `current namespace` variable to shorting the line length done - possibly there is a better way to do this

[the remark this answers is not in the document]
Comment (pablobm) [reply lib/generators/administrate/routes/routes_generator.rb:72]: OK, you are right because the specs fail without those lines :sweat_smile: 

I'm starting to see... The assignment, the array, and the first of the two calls to `add_namespace_to_hash` are unnecessary, right? I think they are red herrings and were throwing me off:

```ruby
      def add_namespace_to_hash(split_path, current_state)
        if split_path.length > 1
           current_namespace = current_state[split_path[0]]
           add_resources_to_hash(split_path[1..], current_namespace[1])
        end
        current_state
      end
```

[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb:72]: so the `add_namespace_to_hash` deals with the next level of nested resources while the `add_resources_to_hash` adds resources into the current namespace level.

[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb:72]: I've not thought about whether the methods need to be called as assignment in the array positions... (so may well work without that) I do think it needs to self call the `add_namespace_to_hash` though to capture the nested namespaces

File: lib/generators/administrate/routes/templates/routes.rb.erb (hunk 11 above)
[the remark this answers is not in the document]
Comment (RaeRachael) [reply lib/generators/administrate/routes/templates/routes.rb.erb:2]: Done

File: lib/generators/administrate/routes/routes_generator.rb (hunk 10 above)
Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb:66]: moved into a new helper file - should make putting tests for it into a sensible location (not recreated yet)

Comment (RaeRachael) [reply lib/generators/administrate/routes/routes_generator.rb]: `resources :authors do # NOTE: not `namespace`
  resources :posts
end
# This gives you things like:
#  - /authors
#  - /authors/:id
#  - /authors/:id/posts
#  - /authors/:id/posts/:id`
 
 Only just looking back at this - I am pretty confident that what is currently in place will function but leads to 
 `resources :authors do
  resources :posts
end
resources: authors`

I feel like this is something to think about next weekend

File: lib/generators/administrate/routes/templates/routes.rb.erb [hunk 12]
```diff
@@ -1,5 +1,4 @@
 namespace :<%= namespace %> do
-<% dashboard_resources.each do |resource| %>    resources :<%= resource %>
-<%end%>
-    root to: "<%= dashboard_resources.first %>#index"
-  end
+<%= generate_resource_routes() %>
```
Comment (pablobm) [inline lib/generators/administrate/routes/templates/routes.rb.erb:2]: Nitpick! Can you remove the brackets please? It's a style thing; you don't normally see empty brackets in Ruby code.

```suggestion
<%= generate_resource_routes %>
```

File: lib/generators/administrate/routes/routes_generator.rb [hunk 13]
```diff
@@ -6,12 +6,19 @@
...
 module Administrate
   module Generators
     class RoutesGenerator < Rails::Generators::Base
       include Administrate::GeneratorHelpers
+      include Administrate::NamespacingHelpers
+      INVALID_DATABASE_MODELS_LIST = [
+        "ActiveRecord::SchemaMigration",
+        "ActiveRecord::InternalMetadata",
+        "primary::SchemaMigration",
+      ]
```
Comment (pablobm) [inline lib/generators/administrate/routes/routes_generator.rb:21]: As per Hound:

```suggestion
      ].freeze
```

File: lib/administrate/namespacing_helpers.rb [hunk 14]
```diff
@@ -0,0 +1,65 @@
+module Administrate
+  module NamespacingHelpers
```
Comment (pablobm) [inline lib/administrate/namespacing_helpers.rb:2]: I think there isn't value in putting these in a mixin, as this code is only referenced from one place: `RoutesGenerator`. If I'm reading the generator, then I find myself wondering "where is generate_resource_routes? Oh, it's in a mixin, let's go there". In fact there's one additional jump, as my search for this code started at the template `routes.rb.erb`. If we were trying to be DRY it could be worth it, but in our case there's less reason for it.

Also, of all these methods, the only one that is used by other code is `generate_resource_routes`. The rest are only used internally and should be private, so that the reader doesn't have to wonder what they are for.

Following on what I said at https://github.com/thoughtbot/administrate/pull/1812/files#r540295544 about factoring code out into a class: that would be a good reason to separate the code. We could this complex, recursive code somewhere that it can be unit-tested, so that we can make sure that every case is covered without having to make our example app any more complex.

**Review by pablobm — CHANGES_REQUESTED** [review]
The build is failing, but I think only because of the security check on dependencies. If you rebase this on top of the latest master, it should work.

File: lib/generators/administrate/routes/templates/routes.rb.erb (hunk 12 above)
Comment (RaeRachael) [reply lib/generators/administrate/routes/templates/routes.rb.erb:2]: will fix, my ruby styling will most likely get worse over time now

File: lib/administrate/namespacing_helpers.rb (hunk 14 above)
Comment (RaeRachael) [reply lib/administrate/namespacing_helpers.rb:2]: I'll look at this more on the weekend - but if I understand the suggestion is to replace the module with a class allowing just he `generate_resource_routes` to be public?

Comment (pablobm) [reply lib/administrate/namespacing_helpers.rb:2]: Not just these methods, but rather all the code that takes part in this complicated calculation of the routes (I'm commenting in passing now, so I'm not sure what else is involved). I linked the wrong comment earlier, I meant https://github.com/thoughtbot/administrate/pull/1812#discussion_r540294128, which includes an example.

**Comment by pablobm** [comment]
Fixes https://github.com/thoughtbot/administrate/issues/1907

File: lib/administrate/namespacing_helpers.rb (hunk 14 above)
Comment (RaeRachael) [reply lib/administrate/namespacing_helpers.rb:2]: I am struggling to see how it works as a class - probably as its part of a larger issue.. As we would want to call a function of that class, `generate_resource_routes`, from a erb file generated by the routes_generator... (is this possible?)
